cps/kni.c

```diff
@@ -668,6 +668,230 @@ rd_modroute(const struct nlmsghdr *req, struct route_update *update)
+		struct gk_fib *fib = &ltbl->fib_tbl[re4->next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm_rule_iterate(&state, &re4);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state.depth, &gw_addr);
+
+		/* Add address and gateway. */
+		mnl_attr_put_u32(reply, RTA_DST, htonl(re4->ip));
+		mnl_attr_put_u32(reply, RTA_GATEWAY, gw_addr->ip.v4.s_addr);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			rd_send_batch(cps_conf, batch, "IPv4",
+				req->nlmsg_seq, req->nlmsg_pid);
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+}
+
+static void
+rd_getroute_ipv6(struct cps_config *cps_conf, struct gk_lpm *ltbl,
+	struct mnl_nlmsg_batch *batch, const struct nlmsghdr *req, int family)
+{
+	struct rte_lpm6_iterator_state state6;
+	struct rte_lpm6_rule re6;
+	int index;
+
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, 0, 0, &state6);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv6 LPM rule iterator state in %s\n",
+			__func__);
+		return;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		struct gk_fib *fib = &ltbl->fib_tbl6[re6.next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm6_rule_iterate(&state6, &re6);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state6.depth, &gw_addr);
+
+		/* Add address and gateway. */
+		mnl_attr_put(reply, RTA_DST,
+			sizeof(struct in6_addr), re6.ip);
+		mnl_attr_put(reply, RTA_GATEWAY,
+			sizeof(struct in6_addr), &gw_addr->ip.v6);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			rd_send_batch(cps_conf, batch, "IPv6",
+				req->nlmsg_seq, req->nlmsg_pid);
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+}
+
+static int
+rd_getroute(const struct nlmsghdr *req)
+{
+	/*
+	 * Buffer length set according to libmnl documentation:
+	 * the buffer that you have to use to store the batch must be
+	 * double of MNL_SOCKET_BUFFER_SIZE to ensure that the last
+	 * message (message N+1) that did not fit into the batch is
+	 * written inside valid memory boundaries.
+	 */
+	char buf[2 * MNL_SOCKET_BUFFER_SIZE];
+	struct mnl_nlmsg_batch *batch;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct gk_lpm *ltbl = &cps_conf->gk->lpm_tbl;
+	int family;
+
+	if (mnl_nlmsg_get_payload_len(req) < sizeof(struct rtgenmsg)) {
+		CPS_LOG(ERR, "Not enough room in CPS GETROUTE message from routing daemon in %s\n",
+			__func__);
+		goto out;
```

Comment:
There are three options for the return value:

```
 * Your callback may return three possible values:
 * 	- MNL_CB_ERROR (<=-1): an error has occurred. Stop callback runqueue.
 * 	- MNL_CB_STOP (=0): stop callback runqueue.
 * 	- MNL_CB_OK (>=1): no problems has occurred.
```

But `MNL_CB_OK` is the only return value which allows the callback runqueue to continue. In essence, we are using `MNL_CB_OK` to just go to the next Netlink message in the queue.

Reply:
Could verify what Bird expects to receive when the first request (e.g. to add a prefix) fails in a two-request batch? Right now, we are returning nothing for errors.

Reply:
It doesn't look like BIRD ever submits multiple requests in the same packet, but I think we should support it in case a routing daemon we use in the future does.

However, this did raise another issue -- it seems that the code that we have now that accepts `RTM_NEWROUTE` and `RTM_DELROUTE` from the routing daemon is incorrect because it does not acknowledge the messages. So BIRD waits for these replies and hangs.

We'll need to do the same for `*GETROUTE`, `*GETADDR`, and `*GETLINK`, and also reply with error messages instead of acknowledgements when needed.

Reply:
Sorry, no acknowledgements for the `GET` types, but we should include error messages if needed.